### PR TITLE
Transitions raycaster to warp backend

### PR DESCRIPTION
**Pull request by pascal-roth**
# Description

Transitions the raycaster to warp

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab/isaaclab/sensors/ray_caster/kernels.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+ALIGNMENT_YAW = wp.constant(1)
+ALIGNMENT_BASE = wp.constant(2)
+
+# Upper-bound ray-cast distance [m] used by camera classes. The actual depth-clipping is applied
+# as a post-process step per data type, so the kernel is always given a large budget.
+CAMERA_RAYCAST_MAX_DIST: float = 1e6
+
+
+@wp.func
+def quat_yaw_only(
```
Comment (pascal-roth) [inline source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:20]: @AntoineRichard this would be one of our math utils, I remember there was the idea to transition all of them to warp. What happened with that idea?

Comment (AntoineRichard) [reply source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:20]: That did not happen in the end, right now we're keeping the torch stuff for compat, there is only so much we can break in a single release x)

Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:20]: could we have another file (math_warp.py) where we place those kernels? otherwise I would now have to introduce that one here and that is not the correct place IMO

File: source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster.py [hunk 2]
```diff
@@ -373,27 +349,59 @@ def _update_buffers_impl(self, env_mask: wp.array):
...
-            if count != 1:  # Mesh is not global, i.e. we have different meshes for each env
+            if count != 1:
                 count = count // self._num_envs
                 pos_w = pos_w.view(self._num_envs, count, 3)
                 ori_w = ori_w.view(self._num_envs, count, 4)
 
-            self._mesh_positions_w[:, mesh_idx : mesh_idx + count] = pos_w
-            self._mesh_orientations_w[:, mesh_idx : mesh_idx + count] = ori_w
+            self._mesh_positions_w_torch[:, mesh_idx : mesh_idx + count] = pos_w
+            self._mesh_orientations_w_torch[:, mesh_idx : mesh_idx + count] = ori_w
```
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster.py:362]: This behavior (pos_w -= pos_offset) is carried over unchanged from the original pre-PR implementation — the same subtraction appeared in the old _update_buffers_impl. This PR refactored it into _update_mesh_transforms without altering the semantics.

You are correct that the mathematically precise formula for a rotated local offset should be:

```python
mesh_world_pos = view_world_pos + quat_rotate(view_world_quat, local_offset)
```

In practice the existing code only differs when a tracked mesh's origin is offset from its physics prim origin *and* the physics prim is rotated — the typical use case (robot visual meshes coincident with their link origin) isn't affected. Fixing this is a separate correctness improvement that should go in its own PR with a test covering non-zero offsets. I've added a note in the commit message to track this.

File: source/isaaclab/isaaclab/sensors/ray_caster/kernels.py (hunk 1 above)
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:20]: Agreed — quat_yaw_only is a math utility and doesn't belong in the ray-caster-specific kernels.py. Moving it to isaaclab/utils/warp/ (e.g. as a new math_warp.py or alongside existing warp utilities) is the right long-term home. However, moving it is out of scope for this PR since it would add a new public symbol to the utils package, potentially touching unrelated modules. I'll keep it here for now and open a follow-up to relocate it properly.

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py [hunk 3]
```diff
@@ -3,28 +3,69 @@
...
+
+    @property
+    def ray_hits_w(self) -> wp.array | None:
+        """The ray hit positions in the world frame [m].
+
+        Shape is (N, B), dtype ``wp.vec3f``. In torch this resolves to (N, B, 3),
+        where N is the number of sensors, B is the number of rays per sensor.
+        Contains ``inf`` for missed hits.
+        """
+        return self._ray_hits_w
```
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py:59]: Fixed in the latest commit (ab81608): public properties pos_w, quat_w, and ray_hits_w now return torch.Tensor again via zero-copy wp.to_torch views of the underlying warp buffers. Internal kernels continue to write directly to the private _pos_w / _quat_w / _ray_hits_w wp.array attributes. No allocations are added — it is the same memory, same device, just a different Python wrapper at the property boundary.

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py [hunk 4]
```diff
@@ -3,28 +3,69 @@
...
+        """
+        self._device = device
+
+        self._pos_w = wp.zeros(num_envs, dtype=wp.vec3f, device=device)
+        self._quat_w = wp.zeros(num_envs, dtype=wp.quatf, device=device)
+        self._ray_hits_w = wp.zeros((num_envs, num_rays), dtype=wp.vec3f, device=device)
+
+        self._pos_w_torch = wp.to_torch(self._pos_w)
+        self._quat_w_torch = wp.to_torch(self._quat_w)
+        self._ray_hits_w_torch = wp.to_torch(self._ray_hits_w)
```
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py:77]: Resolved by the API fix in the same commit: the public properties now return these torch views, so _pos_w_torch, _quat_w_torch, and _ray_hits_w_torch are no longer dead code.

File: source/isaaclab/isaaclab/sensors/ray_caster/kernels.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        yaw_q = quat_yaw_only(combined_quat)
+        rot_drift = wp.quat_rotate(yaw_q, rcd)
+        pos_drifted = wp.vec3f(combined_pos[0] + rot_drift[0], combined_pos[1] + rot_drift[1], combined_pos[2])
+        ray_starts_w[env_id, ray_id] = wp.quat_rotate(yaw_q, local_start) + pos_drifted
+        ray_directions_w[env_id, ray_id] = local_dir
+    else:
+        rot_drift = wp.quat_rotate(combined_quat, rcd)
+        pos_drifted = wp.vec3f(combined_pos[0] + rot_drift[0], combined_pos[1] + rot_drift[1], combined_pos[2])
+        ray_starts_w[env_id, ray_id] = wp.quat_rotate(combined_quat, local_start) + pos_drifted
+        ray_directions_w[env_id, ray_id] = wp.quat_rotate(combined_quat, local_dir)
```
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:120]: Added a four-line comment in the latest commit explaining the intent: ray STARTS are rotated by the yaw-only quaternion (so the scan footprint follows body heading), but ray DIRECTIONS are intentionally kept in world frame so a downward-pointing scanner continues to fire straight down regardless of pitch/roll.

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py [hunk 6]
```diff
@@ -158,168 +181,186 @@ def _initialize_warp_meshes(self):
...
 
-        self._ray_starts_w[env_ids] = ray_starts_w
-        self._ray_directions_w[env_ids] = ray_directions_w
+        # Data buffers
+        self._data.create_buffers(self._view.count, self.num_rays, self._device)
+
+        # Dummy distance/normal buffers required by the merged raycast_mesh_masked_kernel signature
+        # (RayCaster does not use distance or normals; return_distance=0 and return_normal=0 are passed)
+        self._dummy_ray_distance = wp.empty((1, 1), dtype=wp.float32, device=self._device)
+        self._dummy_ray_normal = wp.empty((1, 1), dtype=wp.vec3f, device=self._device)
```
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py]: Expanded the comment in the latest commit to document the safety contract explicitly: writes to these buffers are gated on return_distance==1 / return_normal==1, both of which are always passed as 0 from RayCaster. The comment now also warns that any future caller enabling those flags must resize the buffers to (num_envs, num_rays) first.

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py:59]: Correction to my earlier reply: the wp.array return is intentional. The goal of this PR is for downstream warp kernels to consume sensor data without a wp.from_torch round-trip, which is only possible if the properties return wp.array directly. The CHANGELOG (4.5.28) now includes an explicit breaking-change note with migration guidance: callers must replace `sensor.data.ray_hits_w` with `wp.to_torch(sensor.data.ray_hits_w)` when a torch.Tensor is needed. The private _*_torch attributes are kept alive as named instance references to prevent GC of the underlying warp memory, but are not part of the public API.

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pascal-roth) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py:77]: Correction to my earlier reply: the _*_torch attributes are intentionally not surfaced by the public properties. They are kept as named instance attributes solely to hold strong references to the zero-copy torch views, preventing Python from GC-ing the underlying warp memory. They are not dead code — without them the warp buffers could be freed prematurely.

**Comment by AntoineRichard** [comment]
# PR Review: Transitions raycaster to warp backend

**Branch:** `pascal/ray-caster-warp` rebased onto latest `develop`
**Scope:** 21 files, +1660/−378 lines

Overall this is a well-executed migration. Clean kernel design with `enable_backward=False`, proper `device=self._device` on every `wp.launch()`, and consistent masked-update pattern. Pre-commit passes cleanly. The test additions are solid (alignment modes, depth clipping independence, sentinel patterns, `quat_yaw_only` regression). Below are the findings grouped by severity.

---

## Critical Issues (4)

### C1: Breaking API — `raycast_dynamic_meshes_kernel` signature changed
**`source/isaaclab/isaaclab/utils/warp/kernels.py:219`**

A new `env_mask` parameter was prepended to `raycast_dynamic_meshes_kernel`. Any user calling the kernel directly via `wp.launch(raycast_dynamic_meshes_kernel, ...)` will silently get wrong results because all arguments shift by one position. Warp kernels can't have optional parameters.

**Suggestion:** Create a new `raycast_dynamic_meshes_masked_kernel` (matching the approach taken for the single-mesh variant `raycast_mesh_masked_kernel` in this same PR) and leave the original signature unchanged. Update internal callers to use the new kernel.

### C2: Breaking API — `RayCasterData` properties return `wp.array` instead of `torch.Tensor`
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py`**

`RayCasterData.pos_w`, `.quat_w`, and `.ray_hits_w` now return `wp.array`. Any downstream code doing `sensor.data.ray_hits_w[:, 2]` will break. The built-in MDP functions were updated, but user code will not be.

Per AGENTS.md, breaking changes require a deprecation first, and Changed/Removed entries need migration guidance. The CHANGELOG entry should at minimum include something like:

> Data properties `pos_w`, `quat_w`, and `ray_hits_w` on `RayCasterData` now return `wp.array` instead of `torch.Tensor`. Use `wp.to_torch(sensor.data.pos_w)` to obtain a PyTorch tensor.

### C3: `RayCasterCamera._debug_vis_callback` passes inf values to visualizer
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_camera.py:440-445`**

The parent `RayCaster._debug_vis_callback` (lines 359-370) filters out `inf` values and guards against zero valid points. The camera subclass does neither — it passes `self.ray_hits_w.view(-1, 3)` directly to the visualizer, which will contain `inf` for missed rays.

### C4: `_apply_depth_clipping` is a silent no-op for unrecognized config values
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_camera.py:451-475`**

The `if/elif` chain handles `"max"` and `"zero"` but any other string (e.g., typo `"Max"`, `"clip"`) silently does nothing. `Literal["max", "zero", "none"]` is not enforced at runtime by `@configclass`. Add an `else: raise ValueError(...)` branch.

---

## Important Issues (11)

### I1: Docstring says "z-component" but code reads x-component
**`source/isaaclab/isaaclab/sensors/ray_caster/kernels.py:203`**

`compute_distance_to_image_plane_masked_kernel` docstring: "This equals the z-component of the hit vector in the camera frame." Code at line 226 reads `disp_cam[0]` — the **x-component**. The inline comment at line 225 correctly says "x-component is the forward (depth) axis."

### I2: Docstrings reference nonexistent `return_distance` parameter
**`source/isaaclab/isaaclab/utils/warp/kernels.py:259` and `:176`**

Both `raycast_dynamic_meshes_kernel` and `raycast_static_meshes_kernel` document `ray_distance` as: "if `return_distance` is True. Otherwise, this array is not used." Neither kernel has a `return_distance` parameter — `ray_distance` is always written via `wp.atomic_min`.

### I3: Config docstring claims `nan` for missed rays but actual value is `+/-inf`
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_camera_cfg.py:57`**

For `depth_clipping_behavior == "none"`, the docstring says `distance_to_image_plane` returns `nan`. The kernel computes `inf * direction_component` which produces `+/-inf`, not `nan`. Users checking `torch.isnan()` as a miss sentinel will fail to detect missed rays.

### I4: `XformPrimView` quaternion convention risk
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py:273-276`**

`XformPrimView.get_world_poses()` returns `(w,x,y,z)` on the USD fallback path but `(x,y,z,w)` on the Fabric GPU path. The code concatenates `[pos, quat]` and views as `wp.transformf` (which expects `(x,y,z,w)`). This works on GPU/Fabric but would produce wrong orientations on CPU-only/no-Fabric. At minimum, add a comment documenting the assumption.

### I5–I6: Dead code — unused torch views
**`ray_caster_data.py:23-25,69-71`** and **`ray_caster.py:256-257`**

- `_pos_w_torch`, `_quat_w_torch`, `_ray_hits_w_torch` in `RayCasterData` — created but never read.
- `_ray_starts_w_torch`, `_ray_directions_w_torch` in base `RayCaster._initialize_rays_impl` — only the multi-mesh camera subclass creates its own versions.

### I7: Per-step `wp.to_torch()` calls instead of cached views
**`ray_caster_camera.py:359-360`**, **`multi_mesh_ray_caster_camera.py:276,282,286,290`**

`wp.to_torch()` is called every step for `_pos_w_wp`, `_quat_w_wp`, `_distance_to_image_plane_wp`, `_ray_distance`, `_ray_normal_w`. While zero-copy, it creates a new Python wrapper each time. The existing warp migration pattern (contact sensor, rigid object) caches torch views at init.

### I8: No guard for invalidated mesh views after simulation reset
**`source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster.py:337-342`**

After `_invalidate_initialize_callback` sets `_view = None`, `MultiMeshRayCaster.mesh_views` (class-level dict) is not cleared. The next `_update_mesh_transforms` call would produce an unhelpful `AttributeError`.

### I9: No test for `raycast_dynamic_meshes_kernel` env_mask behavior
**`source/isaaclab/isaaclab/utils/warp/kernels.py:219`**

Only the single-mesh `raycast_mesh_masked_kernel` is tested for env_mask. The multi-mesh kernel has fundamentally different logic (`atomic_min` for closest-hit, mesh pose transforms) and needs independent env_mask verification.

### I10: No integration tests for `MultiMeshRayCaster` sensor class
**`source/isaaclab/isaaclab/sensors/ray_caster/multi_mesh_ray_caster.py`**

146 lines changed with an entirely new `_update_buffers_impl`. Existing tests only cover the low-level `raycast_dynamic_meshes` function, not the sensor class itself.

### I11: Drift buffers use `wp.from_torch(.contiguous())` instead of warp-primary pattern
**`source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_camera.py:299-300`**

Comment says "zero-copy" but `.contiguous()` on a non-contiguous tensor would silently create a copy, breaking the shared-memory assumption. The parent `RayCaster` correctly uses warp-primary buffers with torch views. The camera subclass should follow the same pattern.

---

## Suggestions

- **S1:** Extract `_env_mask_to_indices(env_mask)` helper — the `wp.to_torch(env_mask).nonzero(...).squeeze(-1)` pattern is repeated 5 times across the sensor classes.
- **S2:** Simplify `_apply_depth_clipping` to a single branch with computed `fill_val` instead of duplicated `wp.launch` calls.
- **S3:** Standardize dummy buffer naming (`_dummy_ray_distance` vs `_dummy_normal_w` vs `_dummy_face_id_w`).
- **S4:** Rename `fill_vec3_inf_kernel` → `fill_vec3_2d_masked_kernel` — name implies inf-only but parameter allows any value.
- **S5:** Document mixed-type API on `MultiMeshRayCasterData` — parent docstring says "all properties return wp.array" but `ray_mesh_ids` is `torch.Tensor`.
- **S6:** Add ray-miss edge case test for `raycast_mesh_masked_kernel` (rays pointing away from mesh).
- **S7:** Add explicit type-contract assertions (`isinstance(sensor.data.ray_hits_w, wp.array)`).
- **S8:** Add `compute_distance_to_image_plane_masked_kernel` unit test with known camera orientation.

---

## Strengths

- Clean kernel design with proper masking support and `enable_backward=False`
- Thorough kernel-level tests for `raycast_mesh_masked_kernel` with all flag combinations
- Strong alignment mode coverage (world/yaw/base) with geometrically precise expected values
- Good end-to-end regression coverage via `test_output_equal_to_usdcamera`
- Well-documented MRO bypass in `MultiMeshRayCasterCamera._initialize_rays_impl`
- Depth clipping independence test verifying one output type's clipping doesn't corrupt the other
- Pre-commit passes cleanly

**Comment by kellyguo11** [comment]
This will likely conflict with https://github.com/isaac-sim/IsaacLab/pull/5179/changes. we should coordinate which one gets in first.

**Comment by pascal-roth** [comment]
@AntoineRichard and @kellyguo11 -- should be good for another review

**Comment by pascal-roth** [comment]
> This will likely conflict with https://github.com/isaac-sim/IsaacLab/pull/5179/changes. we should coordinate which one gets in first.

yes, that might clash, whatever is more ready should get in first

**Comment by AntoineRichard** [comment]
I'd merge this one first. @pascal-roth I made some touch-ups, lmk when you had time to look at them!

File: source/isaaclab/docs/CHANGELOG.rst [hunk 7]
```diff
@@ -83,6 +124,7 @@ Removed
   Use ``isaaclab_physx.sim.spawners.materials`` instead.
 * Removed :class:`DeformableObjectSpawnerCfg` from ``isaaclab.sim.spawners``. Use
   ``isaaclab_physx.sim.spawners.spawner_cfg.DeformableObjectSpawnerCfg`` instead.
+>>>>>>> origin/develop
```
Comment (kellyguo11) [inline source/isaaclab/docs/CHANGELOG.rst]: lingering line from merge?

**Review by kellyguo11 — APPROVED** [review]

**Comment by kellyguo11** [comment]
could you also double check on the test failures? I though our CI should be more green now.

File: source/isaaclab/docs/CHANGELOG.rst (hunk 7 above)
Comment (AntoineRichard) [reply source/isaaclab/docs/CHANGELOG.rst]: Weird that the pre-commit check didn't catch it....

**Comment by pascal-roth** [comment]
cannot reproduce any of the test failures, will retrigger the tests

**Comment by AntoineRichard** [comment]
@pascal-roth Look at the last changes. Looks like it was a garbage collection issue on CI....

**Comment by pascal-roth** [comment]
amazing catch! 

then we just have to document this as another breaking change as the classvar is changing
